pkg/dicomio/reader.go

```diff
@@ -67,7 +67,12 @@ type Reader interface {
 	// SetCodingSystem sets the charset.CodingSystem to be used when ReadString
 	// is called.
 	SetCodingSystem(cs charset.CodingSystem)
+	// ByteOrder returns the byte order of the reader (endianness)
 	ByteOrder() binary.ByteOrder
+	// SkipPixelData returns whether or not the reader will skip pixel data
+	SkipPixelData() bool
```

Comment:
I'm not sure if I like the model of adding this option to the dicomio.Reader for the sole purpose of passing it along to where it's needed later in read.go (e.g the dicomio.Reader is just serving to carry this option to readNativeFrames--the dicomio.Reader itself is not concerned with whether SkipPixelData is true or not).

I'll need to think about a better way of doing this (we can leave that to a later refactor), but I wonder if for now we should just pass a skipPixelData bool to `readElement` from the parsing logic in parse.go and pass it through to readPixelData. 

WDYT?

Reply:
Generally, I like to avoid constructors that have a long list of fixed position arguments, but I'm happy to remove the setter and add a fourth parameter to the constructor. In the beginning, I probably would've created a Config or Opts struct for the constructor that could be extended over time without adding more arguments to the constructor.

Reply:
Although as I'm making the requested change now, I realize this is a backwards-incompat change to `NewReader`. You probably don't want that? That's another advantage of having an extensible struct arg for the constructor.

Reply:
I think my more important point is that I don't think this value belongs in dicomio.Reader in the first place, largely because it has nothing to do with any logic the dicomio.Reader is currently responsible for if I understand correctly. e.g my comment: 

> I wonder if for now we should just pass a skipPixelData bool to readElement from the parsing logic in parse.go and pass it through to readPixelData.

Longer term, it may be better to have something like a readOptions variadic parameters, or just pass along the ParseOption set.

Reply:
Oh I thought your first comment suggested I add a new arg to NewReader()?

Reply:
Sorry, if you look at the file _this_ comment is the first thread, followed by the other one--sorry for the confusion. What I'm trying to convey is:

1. Should this even be in the dicomio.Reader?
2. If yes, then it should be part of NewReader.

Let me know what you think!

Reply:
To me, I makes sense for it to be an option in `dicomio.Reader` because it determines how the reader reads the DICOM data. The reader is already passed around through the global `read*` functions, too. By having the option in the reader, you avoid having to pass this new option (or a list of variadic options) through all of those read functions.

In summary:

1. If you're not concerned with backwards-incompat changes, I agree that adding another arg to `NewReader` is better than adding the setter. (Also, maybe an extensible config/opts struct for the constructor is better moving forward?)
3. I do think the option still makes sense to live within `dicomio.Reader`, regardless of how the option gets set.

Reply:
>To me, it makes sense for it to be an option in dicomio.Reader because it determines how the reader reads the DICOM data.

I don't think this is true, at least of the dicomio.Reader (which is my whole issue with this approach). The job of the dicomio.Reader is to perform low-level binary read options over the dicom data. The option you've passed into the dicomio.Reader is not used anywhere inside the dicomio.Reader struct or methods.

The option is used in the read* methods, which are distinct and seperate from the dicomio.Reader struct itself. The dicomio.Reader struct is _used_ by the read* methods in read.go for sure. 

I guess what I don't like about this is that it seems we are tagging on an extra property to dicomio.Reader (that the dicomio.Reader itself doesn't use/care about) just because the dicomio.Reader struct is passed around to some of the read* methods elsewhere so it's convinient. 

Instead, what I think should be done is the read* methods should directly take a set of options, maybe in the form of a specific configuration struct (whose sole role it is to hold a collection of config options) or maybe a set of options similar to ParseOption... if that makes sense. I'd rather not collect these things into dicomio.Reader and just use dicomio.Reader to "hold onto" stuff we want to pass around elsewhere. 

What do you think?

Reply:
Actually, there's an interesting approach in #245. There, the whole Parse Option Set is added to the Dataset. At first it seems a bit similar to this, but what I like about it is that anyone who holds a Dataset struct can quickly see what ParseOptions it was constructed with. This seems useful in of itself as a property of the output Dataset. 

It happens to also be useful for passing around the option set to the read* methods, though I still feel something is a bit odd with it. But maybe that's an okay way to go for now, and I'll take another pass at some refactors to option passing later if needed. 

WDYT about doing a similar thing here?

Reply:
1. The dataset should just be the dataset. Parsing options should be the concern of the parser, not the underlying dataset.
2. Read options should be the concern of the reader.
3. The reader interface should offer and implement the read function via receivers, much like the various reader interfaces in the stdlib.
4. When you combine 2 & 3, it all comes together nicely, I think.

Reply:
> The dataset should just be the dataset. Parsing options should be the concern of the parser, not the underlying dataset.

That is the ethos behind my objection in this comment on this PR, however there is one very very appealing feature in the case of the Dataset. The Dataset is an output, and any caller code handling the Dataset may find it useful to know what parameters (options) it was generated with. But like I said, other than that I still don't love it and would like a different long term solution along the lines of my first comment in this thread.

> The reader interface should offer and implement the read function via receivers, much like the various reader interfaces in the stdlib.

I think there's still some conflation of multiple things going on. There's the `dicomio.Reader` struct, which is responsible for reading dicom _primitives_ from the underlying byte stream. Then there's the separate code in `read.go` that uses this primitive reader to do much more complicated things (this "primitive" is similar to the "primitive" io.Reader that allows one to read bytes). The primitive reader has no need for nor any concern for the SkipPixelData option you added here. In this PR, it's attached to the dicomio.Reader for the sole purpose of being "passed around" to other functions. This is my concern. 

The read.go functions do way more than what a primitive reader might (and note they're unexported). Semantics of "read" and "parse" aside this is what happens today (one may say the functions in read.go are doing "parsing" but imo it's a bit immaterial to this discussion. maybe one day we'll refactor these internal names, but not critical).

As for a short term solution, like I indicated in my prior comment https://github.com/suyashkumar/dicom/pull/243#discussion_r964285470, simply passing through a set of options (maybe in the form of a config struct) to functions in the read.go file seems acceptable to me.

Whether we should attach the read.go functions into a struct to me isn't a big deal either way. It's all an internal impl detail since they're unexported, and the only upside is saving maybe 1-2 parameters in some functions. for now, I'd prefer just adding a config struct that holds onto some of these options.

Reply:
> The read.go functions do way more than what a primitive reader might (and note they're unexported). Semantics of "read" and "parse" aside this is what happens today (one may say the functions in read.go are doing "parsing" but imo it's a bit immaterial to this discussion. maybe one day we'll refactor these internal names, but not critical).

I'd be a proponent of combining the parser and reader into a single interface to simplify everything here. From there, the interface could demand that a `Read()` function be implemented. The implementer then has free reign on how to read/parse any of the DICOM data, primitives and all. The library could offer a default reader implementation via receiver functions for folks that don't want to create their own reader (probably the large majority of library users).

Having said all of that, in the interest of avoiding massive scope creep for such a small addition, I can move forward with what you prefer. I respect that this is your library.